### ISSUE TITLE
Feature Request: MX40 Pro Support

### ISSUE THREAD

**Issue by TreborKutscher** [bitfocus/companion-module-novastar-controller#27]
Is it possible to add the mx40 pro led controller?

Also support variables  output luminance in nits and color temprature in kelvin would be nice.

**Comment by TreborKutscher** [comment]
[URL Protocols for COEX Controller.pdf](https://github.com/bitfocus/companion-module-novastar-controller/files/10154424/URL.Protocols.for.COEX.Controller.pdf)

I was able to gather some information about the coex api. Maybe this is helpful

**Comment by krisdahl** [comment]
I'm working on the MX40 support now.  The web server port is 8001, which isn't in the documentation.

The functionality will be somewhat limited.  We can blackout, put to normal mode, adjust brightness, switch sources, recall presets, turn on HDR, dynamic boosting, and change color space.

The API is very different from past controllers, so it will be a little odd to implement in the existing module.   It may make more sense to create a Novastar COEX module for the newest controllers.

Also, we can get lists of sources etc. from the API which we can't do from the old controllers, so that would be nice to implement in future.

**Comment by TreborKutscher** [comment]
Nice to know that you start working on mx40pro implementations l.

I also would agree on a Novastar Coex module. Also they now have a half 19inch six port conteoller for coex wich also would fit the coex api.

**Comment by krisdahl** [comment]
Yes the new KU20.  I wrote the module that I think will be compatible with the KU20 (eg pulls the input list directly from the API), but I don't have one to test with so won't be 100% sure it works.  I probably won't specifically advertise that model in the supported devices, but if you set it up as an MX40 it should work the same.

Hoping to have the code finalized and committed early this next week, waiting on a repo and some additional QA.

**Comment by TreborKutscher** [comment]
Hello Krisdahl,

long time no hear. I can now confirm that i have four mx40 in my facility and would be eager to test your module.

**Comment by kman1898** [comment]
Any updates on when this might be implemented?

**Comment by krisdahl** [comment]
@kman1898 I've been working on it but there appears to be some issues with the API.  I'm working with the Novastar tech team to try to figure out what's going on.  I think the new firmware updates caused some issues.

Just for reference which processor do you have?

**Comment by kman1898** [comment]
Understood. Thanks for the info. And the MX40 is what I have access to.

**Comment by markbpx** [comment]
[COEX-Series-Interface-API-User-Manual.pdf](https://github.com/bitfocus/companion-module-novastar-controller/files/12038597/COEX-Series-Interface-API-User-Manual.pdf)

**Comment by markbpx** [comment]
@krisdahl Are there any particular commands which are having issues? I got Blackout, Normal, Freeze and Preset recalls working today using http PUT commands in companion. Using the commands in this document from Nova. I have the Nova tech team visiting our shop tomorrow so I am going to ask them to help me with Brightness as I can’t get that working. 
Is there anything I can do to help with the module?
Cheers for working on it!
/
[COEX-Series-Interface-API-User-Manual.pdf](https://github.com/bitfocus/companion-module-novastar-controller/files/12038651/COEX-Series-Interface-API-User-Manual.pdf)

**Comment by markbpx** [comment]
Forgot to mention I am running latest firmware on MX40 and latest version of companion

**Comment by krisdahl** [comment]
@markbpx I got it all working now, and I did begin working on a node module for the MX/COEX processors.

Brightness is on a per-cabinet basis (you have to tell it which cabinet IDs you want to adjust brightness). I'm working on having a function that will just send to all connected cabinets, should be pretty easy.  Probably not an easy way to just do that with REST bulk commands in companion.

Here is the repo for the library I'm working on--I'll try to get some additional progress today and tomorrow maybe get a beta companion module up.

https://github.com/atomicinfotech/novastar-coex

**Comment by markbpx** [comment]
Thanks @krisdahl 
If you replace the cabinet ID string with 16777215, it will send the command to all cabinets. 
eg: {"idList": [ 16777215 ], "ratio": 0.1, "nit": 1000} will send brightness 10% to all connected cabinets. 
ID 16777215 can also be be used in any API command that requires cabinet ID, and will send the command to all connected cabinets. Just had NovaStarDavid here and helped me figure that out. Hope this helps!

**Comment by krisdahl** [comment]
Okay that helps a lot. Thanks!

**Comment by markbpx** [comment]
@krisdahl Awesome work getting the module up! Thanks! Do you know how I can go about testing it? I tried loading it into the companion dev but doesn't seem to show up. I am pretty new to it so might be doing something wrong!

**Comment by krisdahl** [comment]
Thanks @markbpx , I'm hoping to finish up bulk of it today.

Are you using version 3.0+?   Make sure you set your module path like in step #5 here.

[Companion Module Path Instructions](https://github.com/bitfocus/companion-module-base/wiki)

At this point I only have a couple of the commands integrated, but it is going fast now that I finished the first version of the novastar-coex library.

**Comment by krisdahl** [comment]
@markbpx and @TreborKutscher I did get a version up and running and tested.  Let me know how it works for you.  

https://github.com/bitfocus/companion-module-novastar-coex
